R/pivot_quantiles.R

```diff
@@ -0,0 +1,160 @@
+  cols <- validate_pivot_quantiles(.data, ...)
+  .data <- .data %>%
+    dplyr::mutate(dplyr::across(tidyselect::all_of(cols), nested_quantiles))
+  if (length(cols) > 1L) {
+    lengths_check <- .data %>%
+      dplyr::transmute(dplyr::across(
+        tidyselect::all_of(cols),
+        ~ map_int(.x, vctrs::vec_size)
+      )) %>%
+      as.matrix() %>%
+      apply(1, function(x) dplyr::n_distinct(x) == 1L) %>%
+      all()
+    if (lengths_check) {
+      .data <- tidyr::unnest(.data, tidyselect::all_of(cols), names_sep = "_")
+    } else {
+      if (.ignore_length_check) {
+        for (col in cols) {
+          .data <- .data %>%
+            tidyr::unnest(tidyselect::all_of(col), names_sep = "_")
+        }
+      } else {
+        cli::cli_abort(c(
+          "Some selected columns contain different numbers of quantiles.",
+          "The result would be a {.emph very} long {.cls tibble}.",
+          "To do this anyway, rerun with `.ignore_length_check = TRUE`."
+        ))
+      }
+    }
+  } else {
+    .data <- .data %>% tidyr::unnest(tidyselect::all_of(cols))
+  }
+  .data
+}
+
+#' Pivot columns containing `dist_quantile` wider
+#'
+#' Any selected columns that contain `dist_quantiles` will be "widened" with
+#' the "taus" (quantile) serving as names and the values in the data frame.
+#' When pivoting multiple columns, the original column name will be used as
+#' a prefix.
+#'
+#' @param .data A data frame, or a data frame extension such as a tibble or
+#'   epi_df.
+#' @param ... <[`tidy-select`][dplyr::dplyr_tidy_select]> One or more unquoted
+#'   expressions separated by commas. Variable names can be used as if they
+#'   were positions in the data frame, so expressions like `x:y` can
+#'   be used to select a range of variables.
+#'
+#' @return An object of the same class as `.data`
+#' @export
+#'
+#' @examples
+#' d1 <- c(dist_quantiles(1:3, 1:3 / 4), dist_quantiles(2:4, 1:3 / 4))
+#' d2 <- c(dist_quantiles(2:4, 2:4 / 5), dist_quantiles(3:5, 2:4 / 5))
+#' tib <- tibble::tibble(g = c("a", "b"), d1 = d1, d2 = d2)
+#'
+#' pivot_quantiles_wider(tib, c("d1", "d2"))
+#' pivot_quantiles_wider(tib, tidyselect::starts_with("d"))
+#' pivot_quantiles_wider(tib, d2)
+pivot_quantiles_wider <- function(.data, ...) {
+  cols <- validate_pivot_quantiles(.data, ...)
+  .data <- .data %>%
+    dplyr::mutate(dplyr::across(tidyselect::all_of(cols), nested_quantiles))
+  checks <- map_lgl(cols, ~ diff(range(vctrs::list_sizes(.data[[.x]]))) == 0L)
+  if (!all(checks)) {
+    nms <- cols[!checks]
+    cli::cli_abort(
+      c("Quantiles must be the same length and have the same set of taus.",
+        i = "Check failed for variables(s) {.var {nms}}."
+      )
+    )
+  }
+  if (length(cols) > 1L) {
+    for (col in cols) {
+      .data <- .data %>%
+        tidyr::unnest(tidyselect::all_of(col)) %>%
+        tidyr::pivot_wider(
+          names_from = "tau", values_from = "q",
```

Comment:
thought: These names were updated in https://github.com/cmu-delphi/epipredict/pull/243 but since the functions got moved around here vs that branch, merging them will probably get hairy. The https://github.com/cmu-delphi/epipredict/pull/243 changes might not percolate to this new file, so reminder to check.

Reply:
Definitely hairy, unfortunately.